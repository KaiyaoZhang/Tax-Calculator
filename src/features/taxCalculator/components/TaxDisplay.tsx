import { JSX, useMemo } from "react";
import { Table, Typography } from "antd";
import type { ColumnsType } from "antd/es/table";
import { useTranslation } from "react-i18next";
import { TaxBracketsType, BandTaxTableRowType } from "../types";

interface TaxDisplayProps {
  taxBrackets: TaxBracketsType;
  annualIncome: string;
}

const TaxDisplay = ({
  taxBrackets,
  annualIncome,
}: TaxDisplayProps): JSX.Element | null => {
  const { Title } = Typography;
  const { t } = useTranslation();

  const columns: ColumnsType<BandTaxTableRowType> = useMemo(
    () => [
      {
        title: t("table_col_income_range"),
        dataIndex: "range",
        key: "range",
      },
      {
        title: t("table_col_tax_rate"),
        dataIndex: "rate",
        key: "rate",
      },
      {
        title: t("table_col_tax_owed"),
        dataIndex: "tax",
        key: "tax",
      },
    ],
    [t]
  );

  const currencyFormatter = useMemo(
    () =>
      new Intl.NumberFormat("en-CA", {
        style: "currency",
        currency: "CAD",
      }),
    []
  );

  const formatCurrency = (value: number): string =>
    currencyFormatter.format(value);

  // Calculate the tax band details
  const calculatedTaxDetails = useMemo(() => {
    let remainingSalary = Number(annualIncome);
    let taxAcc = 0;
    const bandBreakdown: BandTaxTableRowType[] = [];

    for (let index = 0; index < taxBrackets.tax_brackets.length; index++) {
      const band = taxBrackets.tax_brackets[index];
      if (remainingSalary <= 0) break;

      const bandMin = band.min;
      const bandMax = band.max ?? Infinity;
      const taxableInBand = Math.min(remainingSalary, bandMax - bandMin);

      // Set the tax, rate, and range for each band and format them to use in Antd Table data source
      if (taxableInBand > 0) {
        const taxInBand = taxableInBand * band.rate;
        taxAcc += taxInBand;
        bandBreakdown.push({
          key: `${index}`,
          range: `${formatCurrency(bandMin)} - ${
            bandMax === Infinity ? "âˆž" : formatCurrency(bandMax)
          }`,
          rate: `${(band.rate * 100).toFixed(1)}%`,
          tax: `${formatCurrency(taxInBand)}`,
        });
        remainingSalary -= taxableInBand;
      }
    }

    const effectiveRate = (taxAcc / Number(annualIncome)) * 100;

    return { totalTax: taxAcc, bandDetails: bandBreakdown, effectiveRate };
  }, [annualIncome, taxBrackets]);

  // Early return when annualIncome is empty or smaller than 0
  if (annualIncome === "" || Number(annualIncome) <= 0) {
    return null;
  }

  return (
    <div>
      <Title level={3}>{t("tax_break_down_table")}</Title>
      <Table
        dataSource={calculatedTaxDetails.bandDetails}
        columns={columns}
        pagination={false}
        bordered
      />
      <Title level={3}>
        {t("totalTaxes", {
          totalTaxes: formatCurrency(calculatedTaxDetails.totalTax),
        })}
      </Title>
      <Title level={3}>
        {`${t("effectiveRate", {
          effectiveRate: calculatedTaxDetails.effectiveRate.toFixed(2),
        })}%`}
      </Title>
    </div>
  );
};

export default TaxDisplay;
