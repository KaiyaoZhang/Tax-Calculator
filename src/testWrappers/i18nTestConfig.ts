import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import enTranslation from "../../public/locales/en/translation.json";
import frTranslation from "../../public/locales/fr/translation.json";

i18n.use(initReactI18next).init({
  lng: "en", // Default test language
  fallbackLng: "en",
  resources: {
    en: { translation: enTranslation },
    fr: { translation: frTranslation },
  },
  interpolation: { escapeValue: false },
});

export default i18n;
