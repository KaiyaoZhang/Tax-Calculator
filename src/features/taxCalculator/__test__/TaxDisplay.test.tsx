import { render, screen } from "@testing-library/react";
import { I18nextProvider } from "react-i18next";
import i18nTestConfig from "../../../testWrappers/i18nTestConfig";
import TaxDisplay from "../components/TaxDisplay";

const { t } = i18nTestConfig;

jest.mock("../../../apis", () => ({
  getTaxRates: jest.fn(),
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe("Tax display page", () => {
  const mockProps = {
    taxBrackets: {
      tax_brackets: [
        {
          max: 48535,
          min: 0,
          rate: 0.15,
        },
        {
          max: 97069,
          min: 48535,
          rate: 0.205,
        },
        {
          max: 150473,
          min: 97069,
          rate: 0.26,
        },
        {
          max: 214368,
          min: 150473,
          rate: 0.29,
        },
        {
          min: 214368,
          rate: 0.33,
        },
      ],
    },
    annualIncome: "250000",
  };

  it("it should show the proper page when have taxBrackets and annualIncome value", () => {
    const { container } = render(
      <I18nextProvider i18n={i18nTestConfig}>
        <TaxDisplay {...mockProps} />
      </I18nextProvider>
    );

    expect(container).toMatchSnapshot();

    expect(screen.getByText(t("tax_break_down_table"))).toBeInTheDocument();
    expect(screen.getByText(t("table_col_income_range"))).toBeInTheDocument();
    expect(screen.getByText(t("table_col_tax_owed"))).toBeInTheDocument();
    expect(screen.getByText(t("table_col_tax_rate"))).toBeInTheDocument();
  });

  it("it should not show the page if annualIncome doesn't have a value", () => {
    const newMockProps = {
      ...mockProps,
      annualIncome: "",
    };
    render(
      <I18nextProvider i18n={i18nTestConfig}>
        <TaxDisplay {...newMockProps} />
      </I18nextProvider>
    );

    expect(
      screen.queryByText(t("tax_break_down_table"))
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText(t("table_col_income_range"))
    ).not.toBeInTheDocument();
  });
});
