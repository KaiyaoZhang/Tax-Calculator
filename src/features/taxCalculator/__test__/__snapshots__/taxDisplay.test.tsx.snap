// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tax display page it should show the proper page when have taxBrackets and annualIncome value 1`] = `
<div>
  <div>
    <h3
      class="ant-typography css-dev-only-do-not-override-19lec04"
    >
      Tax Breakdown table:
    </h3>
    <div
      class="ant-table-wrapper css-dev-only-do-not-override-19lec04"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-19lec04"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-bordered css-dev-only-do-not-override-19lec04"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup />
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Income Range
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Tax Rate
                      </th>
                      <th
                        class="ant-table-cell"
                        scope="col"
                      >
                        Tax Owed
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="0"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        $0.00 - $48,535.00
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        15.0%
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        $7,280.25
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="1"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        $48,535.00 - $97,069.00
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        20.5%
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        $9,949.47
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="2"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        $97,069.00 - $150,473.00
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        26.0%
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        $13,885.04
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="3"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        $150,473.00 - $214,368.00
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        29.0%
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        $18,529.55
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="4"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        $214,368.00 - âˆž
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        33.0%
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        $11,758.56
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <h3
      class="ant-typography css-dev-only-do-not-override-19lec04"
    >
      Total Taxes: $61,402.87
    </h3>
    <h3
      class="ant-typography css-dev-only-do-not-override-19lec04"
    >
      Effective Rate: 24.56%
    </h3>
  </div>
</div>
`;
